### ISSUE TITLE
Missed MEV blocks when DB is being flushed

### ISSUE THREAD

**Issue by 0xalex88** [prysmaticlabs/prysm#12501]
# 💎 Issue

### Background

Missed MEV blocks when DB is being flushed between GetBeaconBlock and ProposeBeaconBlock

### Description

In the last 4 days we've missed multiple blocks because the ProposeBeaconBlock call was too late into the block and the relay was rejecting the request.

Just as a comparison these are successfully proposed blocks:

```
{"time":"2023-06-05T07:24:12Z","backend":null,"duration":663709,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/DomainData","msg":"gRPC request finished."}
{"time":"2023-06-05T07:24:13Z","backend":null,"duration":1631219526,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/ProposeBeaconBlock","msg":"gRPC request finished."}
{"time":"2023-06-05T07:24:13Z","blockRoot":"0x582d6f13959d","fork":"capella","gasUtilized":0.4119289666666667,"level":"info","logrus_error":"can not add field \"blockNumber\"","msg":"Submitted new block","numAttestations":128,"numDeposits":0,"parentHash":"0x1fa60dc25482","payloadHash":"0xa7c36a64246d","prefix":"validator","pubKey":"0x887be6ce8f4f","slot":6593819}

{"time":"2023-06-05T09:34:00Z","backend":null,"duration":354712189,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/GetBeaconBlock","msg":"gRPC request finished."}
{"time":"2023-06-05T09:34:00Z","backend":null,"duration":722865,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/DomainData","msg":"gRPC request finished."}
{"time":"2023-06-05T09:34:03Z","backend":null,"duration":172118713,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/GetAttestationData","msg":"gRPC request finished."}
{"time":"2023-06-05T09:34:03Z","backend":null,"duration":172278129,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/GetAttestationData","msg":"gRPC request finished."}
{"time":"2023-06-05T09:34:03Z","backend":null,"duration":172407456,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/GetAttestationData","msg":"gRPC request finished."}
{"time":"2023-06-05T09:34:03Z","backend":null,"duration":172200371,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/GetAttestationData","msg":"gRPC request finished."}
{"time":"2023-06-05T09:34:03Z","backend":null,"duration":172272326,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/GetAttestationData","msg":"gRPC request finished."}
{"time":"2023-06-05T09:34:03Z","backend":null,"duration":172193706,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/GetAttestationData","msg":"gRPC request finished."}
{"time":"2023-06-05T09:34:03Z","backend":null,"duration":172288941,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/GetAttestationData","msg":"gRPC request finished."}
{"time":"2023-06-05T09:34:03Z","backend":null,"duration":3025100471,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/ProposeBeaconBlock","msg":"gRPC request finished."}
{"time":"2023-06-05T09:34:03Z","blockRoot":"0x6d0dd02e9952","fork":"capella","gasUtilized":0.6200066,"level":"info","logrus_error":"can not add field \"blockNumber\"","msg":"Submitted new block","numAttestations":128,"numDeposits":0,"parentHash":"0x810109d6bb81","payloadHash":"0x9352d0680041","prefix":"validator","pubKey":"0x993aaec654d5","slot":6594468}
```

and this is one of the failed ones:

```
{"time":"2023-06-06T05:06:35Z","backend":null,"duration":242208864,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/GetBeaconBlock","msg":"gRPC request finished."}
{"time":"2023-06-06T05:06:35Z","backend":null,"duration":4562097,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/DomainData","msg":"gRPC request finished."}
{"time":"2023-06-06T05:06:39Z","backend":null,"duration":4727354,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/GetAttestationData","msg":"gRPC request finished."}
... repeated
{"time":"2023-06-06T05:06:39Z","backend":null,"duration":15232512,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/GetSyncMessageBlockRoot","msg":"gRPC request finished."}
{"time":"2023-06-06T05:06:39Z","backend":null,"duration":5037618,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/DomainData","msg":"gRPC request finished."}
{"time":"2023-06-06T05:06:39Z","backend":null,"duration":4562705,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/SubmitSyncMessage","msg":"gRPC request finished."}
{"time":"2023-06-06T05:06:39Z","blockRoot":"0x8f60606df3c6","level":"info","msg":"Submitted new sync message","prefix":"validator","slot":6600331,"slotStartTime":"2023-06-06T05:06:35Z","timeSinceSlotStart":4036950843,"validatorIndex":564558}
{"time":"2023-06-06T05:06:39Z","level":"debug","msg":"Batched attestation records write interval reached, flushing to DB","numRecords":30,"prefix":"db"}
{"time":"2023-06-06T05:06:42Z","duration":3874464236,"level":"debug","msg":"Successfully flushed batched attestations to DB","prefix":"db"}
{"time":"2023-06-06T05:06:42Z","backend":null,"duration":5851624,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/ProposeAttestation","msg":"gRPC request finished."}
... repeated
{"time":"2023-06-06T05:06:42Z","backend":null,"duration":5684571,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/ProposeAttestation","msg":"gRPC request finished."}
{"time":"2023-06-06T05:06:44Z","backend":null,"duration":2225980675,"level":"debug","method":"/ethereum.eth.v1alpha1.BeaconNodeValidator/ProposeBeaconBlock","msg":"gRPC request finished."}
{"time":"2023-06-06T05:06:44Z","blockSlot":6600331,"error":"rpc error: code = Unknown desc = could not submit blinded block: error posting the SignedBlindedBeaconBlockCapella to the builder api: unsupported error code: 502: did not receive 200 response from API","level":"error","msg":"Failed to propose block","prefix":"validator","pubKey":"0x848c8f4e69d4"}
```

I've noticed the difference is when the ProposeBeaconBlock is called after the database flush (which for some reason instead of taking the usual 30-50ms, takes ~3 seconds). That's the only time the blocks were missed.

The disk is an OVH high speed volume, I've tested it using `fio` and I've got 3k IOPS read and 1k IOPS write, using `dd` in sync mode shows 100MB/s writes, looking at `iotop` cumulative usage seems the validator client doesn't write less than 1MB of data every block. I've tried to look at the historical write performance and most of the time is within 50ms but sometimes like in the missed block above is about 3seconds.

The GetBeaconBlock call (240ms) and the ProposeBeaconBlock call (2.2s because of all the retries) didn't seem too long, it seems to me (and I could very well be wrong about my assumption) that when the signing takes long enough that the attestations kicks in, the flush of the database takes longer and the proposal slashing db check has to wait for the db to flush.

**Comment by 0xalex88** [comment]
Just as an addition, the database seems to be constantly growing and pruning doesn't reduce its size.
We have three instances, they have 3.8G, 873M, 69M databases. First one after 20 minutes of pruning wasn't done, second one took 4-5 minutes and then it started correctly, but didn't decrease its size (I would've expected to shrink up to the 3rd instance size)

**Comment by prestonvanloon** [comment]
@0xalex88 -- Unfortunately, boltdb is not able to release allocated space back to the disk once it has been allocated. It does free up some space for storage and will use that but the total size on disk will only increase and never decrease. When you prune the db, it may "free" 1gb which can be overwritten but the on disk size will remain the same. The only way to truly shrink the disk space footprint is to iterate over the whole db and write to a new database.

See: https://github.com/boltdb/bolt/issues/423

**Comment by 0xalex88** [comment]
@prestonvanloon oh that's good to know! thanks

**Comment by nisdas** [comment]
@0xalex88 I think we have root caused this correctly and solved this in v4.0.7. If this is not the case for you, I will reopen the issue.

**Comment by 0xalex88** [comment]
@nisdas thank you! will do!
